src/widgets.rs

```diff
@@ -227,3 +216,323 @@ pub trait StatefulWidget {
     type State;
     fn render(self, area: Rect, buf: &mut Buffer, state: &mut Self::State);
 }
+
+/// A `WidgetRef` is a trait that allows rendering a widget by reference.
+///
+/// This trait is useful when you want to store a reference to a widget and render it later. It also
+/// allows you to render boxed widgets.
+///
+/// Boxed widgets allow you to store widgets with a type that is not known at compile time. This is
+/// useful when you want to store a collection of widgets with different types. You can then iterate
+/// over the collection and render each widget.
+///
+/// This trait was introduced in Ratatui 0.26.0 and is implemented for all the internal widgets.
+/// Implemetors should prefer to implement this over the `Widget` trait and add an implementation of
+/// `Widget` that calls `WidgetRef::render_ref` where backwards compatibility is required.
+///
+/// A blanket implementation of `Widget` for `&W` where `W` implements `WidgetRef` is provided.
+///
+/// A blanket implementation of `WidgetRef` for `Option<W>` where `W` implements `WidgetRef` is
+/// provided. This is a convenience approach to make it easier to attach child widgets to parent
+/// widgets. It allows you to render an optional widget by reference.
+///
+/// # Examples
+///
+/// ```rust
+/// use ratatui::{prelude::*, widgets::*};
+///
+/// struct Greeting;
+///
+/// struct Farewell;
+///
+/// impl WidgetRef for Greeting {
+///     fn render_ref(&self, area: Rect, buf: &mut Buffer) {
+///         Line::raw("Hello").render(area, buf);
+///     }
+/// }
+///
+/// /// Only needed for backwards compatibility
+/// impl Widget for Greeting {
+///     fn render(self, area: Rect, buf: &mut Buffer) {
+///         self.render_ref(area, buf);
+///     }
+/// }
+///
+/// impl WidgetRef for Farewell {
+///     fn render_ref(&self, area: Rect, buf: &mut Buffer) {
+///         Line::raw("Goodbye").right_aligned().render(area, buf);
+///     }
+/// }
+///
+/// /// Only needed for backwards compatibility
+/// impl Widget for Farewell {
+///     fn render(self, area: Rect, buf: &mut Buffer) {
+///         self.render_ref(area, buf);
+///     }
+/// }
+///
+/// # fn render(area: Rect, buf: &mut Buffer) {
+/// let greeting = Greeting;
+/// let farewell = Farewell;
+///
+/// // these calls do not consume the widgets, so they can be used again later
+/// greeting.render_ref(area, buf);
+/// farewell.render_ref(area, buf);
+///
+/// // a collection of widgets with different types
+/// let widgets: Vec<Box<dyn WidgetRef>> = vec![Box::new(greeting), Box::new(farewell)];
+/// for widget in widgets {
+///     widget.render_ref(area, buf);
+/// }
+/// # }
+/// ```
+pub trait WidgetRef {
+    fn render_ref(&self, area: Rect, buf: &mut Buffer);
+}
```

Comment:
Can we make `WidgetRef` `pub(crate)` only for a version or `unstable`? It'll help understand how this is going to be used, at least for me. 

The docstring also recommends implements use this instead of `Widget`, and I'm worried I have not explored all the facets of this implementation. I just don't understand the implications of this enough to have lots of confidence in this approach.

But I can tell it'll be useful for `Box` and potentially widget lists etc. So if other team members are on board with this, I'm too!

Reply:
Can you expand on why this doesn't work?

```rust
impl Widget for Box<Block<'_>> {
    fn render(self, area: Rect, buf: &mut Buffer) {
        self.as_ref().render(area, buf);
    }
}
```

I added this test:

```rust
    #[test]
    fn box_block() {
        let mut buffer = Buffer::empty(Rect::new(0, 0, 4, 2));
        let block = Box::new(
            Block::default()
                .title("test")
                .title_position(Position::Bottom),
        );
        block.render(buffer.area, &mut buffer);
        assert_buffer_eq!(buffer, Buffer::with_lines(vec!["    ", "test"]));
    }
```

And I was able to compile and run it:

<img width="709" alt="image" src="https://github.com/ratatui-org/ratatui/assets/1813121/645cc848-f17e-4950-bc99-2062e19f6ae1">

Reply:
> Can we make WidgetRef pub(crate) only for a version or unstable? It'll help understand how this is going to be used, at least for me.

Yep that will work. I added this to the release highlights PR, but I can remove it.

> Can you expand on why this doesn't work?

```rust
impl Widget for Box<Block<'_>> {
    fn render(self, area: Rect, buf: &mut Buffer) {
        self.as_ref().render(area, buf);
    }
}
```

Sure, lets start with the following example:

```rust
struct WidgetInABox;

impl Widget for WidgetInABox {
    fn render(self, area: Rect, buf: &mut Buffer) {
        Line::raw("Hello").render(area, buf);
    }
}

impl Widget for &WidgetInABox {
    fn render(self, area: Rect, buf: &mut Buffer) {
        Line::raw("Hello").render(area, buf);
    }
}

impl Widget for Box<WidgetInABox> {
    fn render(self, area: Rect, buf: &mut Buffer) {
        self.as_ref().render(area, buf);
    }
}
```

Remove `impl Widget for &WidgetInABox` and the `self.as_ref().render(area, buf);` line does not compile:

```rust
error[E0507]: cannot move out of a shared reference
   --> src/widgets.rs:555:13
    |
555 |             self.as_ref().render(area, buf);
    |             ^^^^^^^^^^^^^ ----------------- value moved due to this method call
    |             |
    |             move occurs because value has type `WidgetInABox`, which does not implement the `Copy` trait
    |
note: `widgets::Widget::render` takes ownership of the receiver `self`, which moves value
   --> src/widgets.rs:101:15
    |
101 |     fn render(self, area: Rect, buf: &mut Buffer);
    |               ^^^^
```
Ok, so we know that that render call has to be on &WidgetInABox not &Widget.
Adding it back in compiles fine, and the following test passes.
```rust
#[rstest]
fn widget_box_render(mut buf: Buffer) {
    let widget = Box::new(WidgetInABox);
    widget.render(buf.area, &mut buf);
    assert_eq!(buf, Buffer::with_lines(["Hello               "]));
}
```
Adding it back in compiles fine, but it's ultimately useless because this doesn't compli
But that's not actually what we want. Widget there is `Box<WidgetInABox>` we want `Box<dyn Widget>` so that we can insert any widget. The following test fails to compile:
```rust
#[rstest]
fn widget_box_dyn_render(mut buf: Buffer) {
    let widget: Box<dyn Widget> = Box::new(WidgetInABox);
    widget.render(buf.area, &mut buf);
    assert_eq!(buf, Buffer::with_lines(["Hello               "]));
}
```
```rust
error[E0161]: cannot move a value of type `dyn widgets::Widget`
   --> src/widgets.rs:569:9
    |
569 |         widget.render(buf.area, &mut buf);
    |         ^^^^^^ the size of `dyn widgets::Widget` cannot be statically determined
```
The only way that I've managed to come up with something workable is to add a new trait that accepts &self (hence WidgetRef). My level of understanding if definitely expert level on this, so there might be something I'm missing. I meant to post on the user's forum to see if there was some unknown approach, but had other things to do last night. Getting to it now.

Reply:
Rust lang post:
https://users.rust-lang.org/t/calling-a-method-with-a-self-parameter-on-box-dyn-trait/106220

Rust playground simplification of the problem:
https://play.rust-lang.org/?version=stable&mode=debug&edition=2021&gist=93f750557d860ea6a10c5c714e7e81c2

Reply:
Also posted it in our discord:
https://discord.com/channels/1070692720437383208/1070692720437383211/1202850480162607124

And the Rust Programming Langudage Discord:
https://discord.com/channels/273534239310479360/1202852180999151667

Reply:
> Can we make WidgetRef pub(crate) only for a version or unstable? It'll help understand how this is going to be used, at least for me.

I've marked it as unstable - at first I was thinking about pushing back, but I think it makes sense to do that. I'll update the PR title to clarify this point.